### DISCUSSION TITLE
Use incron for automating a few applet triggers

### DISCUSSION THREAD

**Discussion by intelliant01** [ltsp/ltsp#576] [Ideas]
I have used incron to automate a few applet triggers based on changes to ltsp.conf or creation or modification of new users. It is quiet simple.

On Ubuntu 18.04:

```
apt install incron
```
Create the table containing rows of `<path to watch> <event to watch for> <trigger/action>`
```
root@server:~# cat /etc/incron.d/ltsp
/etc/shadow IN_ATTRIB,IN_NO_LOOP ltsp initrd
/etc/group IN_ATTRIB,IN_NO_LOOP ltsp initrd
/etc/ltsp/ltsp.conf IN_ATTRIB,IN_CLOSE_WRITE,IN_NO_LOOP ltsp ipxe
```

In case one wants to extend the use beyond the above, using incron could be tricky. For example, your version of incron may not accept comments in the table file.
Found a great troubleshooting reference herein - https://hackaday.com/2020/10/28/linux-fu-troubleshooting-incron/

@alkisg May be this could become a default ltsp feature.

**Comment by alkisg** [comment]
Indeed it sounds useful for people that forget to run `ltsp initrd` after changes in passwd, group, in /etc/ltsp or in /usr/share/ltsp.

But I won't integrate anything involving `incron` upstream in LTSP, because the proper implementation would be via the LTSP http service:

https://github.com/ltsp/ltsp/issues/147

**Reply by intelliant01** [reply]
Yes, #147 may make the need for incron redundant in the ltsp context. Till then, hopefully it will make life simpler for forgetfuls like me. ;)

**Comment by alkisg** [comment]
See also this: https://github.com/ltsp/ltsp/issues/583#issuecomment-946862297

I.e. same thing as incron, but with a systemd .path unit, so that no resources or dependencies are needed.

**Reply by intelliant01** [reply]
I had studied both and then sided with incron as it has more options, with less files to manage per watch I wish to add. But that also means having to dig a bit deeper as incron can get tricky for starters. It is of utility in other contexts for me as well on the same server.

For watching on user addition or modification and safeguard against a race condition for mass user modifications, I suggest using a script with inotifywait with --timeout and based on the exit value 0, you may loop to watch further or for value 2 execute ltsp initrd. I hope to next work on this.

**Reply by intelliant01** [reply]
Sometimes the mind takes you for an fast ride when all you need is a stroll. ;) 

Incron config: 
```
# cat /etc/incron.d/ltsp
/etc/shadow IN_ATTRIB,IN_NO_LOOP execonchangestop $@ ltsp initrd
/etc/group IN_ATTRIB,IN_NO_LOOP execonchangestop $@ ltsp initrd
```

Simpler than using `inotifywait` that I had suggested earlier, a script `/usr/bin/execonchangestop` -

```
#!/bin/bash

args=("$@")
monitored_file=${args[0]}
shift;
cmd="$*"

while [ $(( $(date +%s) - $(stat -c %Y $monitored_file) )) -lt 10 ]
do
   sleep 1 
done

eval $cmd
```
So incron watches `/etc/shadow`, `/etc/group` for `IN_ATTRIB` event and triggers the script which executes `ltsp initrd` if no changes have happened to the watched file for 10 seconds.

**Comment by alkisg** [comment]
If someone calls adduser in a loop, and modifies /etc/passwd 100 times, won't incron invoke that script 100 times?
That would result in 10 seconds delay, and then 100 `ltsp initrd` invocations...

A systemd.path unit can do the same thing with no external dependencies and it additionally offers rate limiting.

**Reply by intelliant01** [reply]
incron will not execute 100 times because of -

```
IN_NO_LOOP. This symbol disables monitoring events until the current one is completely handled (until its child process exits).
```

I think I had tested this scenario as well. Will do so again when back at it.

**Comment by alkisg** [comment]
That's nice, so incron also supports rate limiting.
Regarding the actual rate, I'll comment in #583.
